lib/tracing-limit/src/lib.rs

```diff
@@ -178,39 +212,125 @@ where
+    }
+
+    #[inline]
+    fn on_close(&self, id: span::Id, ctx: Context<'_, S>) {
+        self.inner.on_close(id, ctx);
+    }
+
+    #[inline]
+    fn on_id_change(&self, old: &span::Id, new: &span::Id, ctx: Context<'_, S>) {
+        self.inner.on_id_change(old, new, ctx);
+    }
+}
+
+impl<S, L> RateLimitedLayer<S, L>
+where
+    S: Subscriber,
+    L: Layer<S>,
+{
+    fn create_event(
+        &self,
+        ctx: &Context<S>,
+        metadata: &'static Metadata<'static>,
+        message: String,
+        rate_limit: u64,
+    ) {
+        let fields = metadata.fields();
+
+        let message = display(message);
+
+        if let Some(message_field) = fields.field("message") {
+            let values = [(&message_field, Some(&message as &dyn Value))];
+
+            let valueset = fields.value_set(&values);
+            let event = Event::new(metadata, &valueset);
+            self.inner.on_event(&event, ctx.clone());
+        } else {
+            let values = [(
+                &fields.field(RATE_LIMIT_SECS_FIELD).unwrap(),
+                Some(&rate_limit as &dyn Value),
+            )];
+
+            let valueset = fields.value_set(&values);
+            let event = Event::new(metadata, &valueset);
+            self.inner.on_event(&event, ctx.clone());
+        }
+    }
+}
+
+#[derive(Debug)]
+struct State {
+    start: Instant,
+    count: AtomicUsize,
+    limit: u64,
+    message: String,
 }
 
 fn is_limited(metadata: &Metadata<'_>) -> bool {
     metadata
         .fields()
         .iter()
-        .any(|f| f.name() == RATE_LIMIT_FIELD)
+        .any(|f| f.name() == RATE_LIMIT_SECS_FIELD)
+}
+
+/// RateLimitedSpanKeys records span keys that we use to rate limit callsites separately by. For
+/// example, if a given trace callsite is called from two different components, then they will be
+/// rate limited separately.
+#[derive(Default)]
+struct RateLimitedSpanKeys {
+    pub component_name: Option<String>,
+}
+
+impl Visit for RateLimitedSpanKeys {
+    fn record_str(&mut self, field: &Field, value: &str) {
+        if field.name() == COMPONENT_NAME_FIELD {
+            self.component_name = Some(value.to_string());
+        }
+    }
+
+    fn record_debug(&mut self, _field: &Field, _value: &dyn fmt::Debug) {}
 }
 
 #[derive(Default)]
 struct LimitVisitor {
     pub limit: Option<usize>,
     pub message: Option<String>,
+    pub key: Option<String>,
 }
 
 impl Visit for LimitVisitor {
     fn record_u64(&mut self, field: &Field, value: u64) {
-        if field.name() == RATE_LIMIT_FIELD {
+        if field.name() == RATE_LIMIT_SECS_FIELD {
             self.limit = Some(value as usize);
         }
     }
 
     fn record_i64(&mut self, field: &Field, value: i64) {
-        if field.name() == RATE_LIMIT_FIELD {
+        if field.name() == RATE_LIMIT_SECS_FIELD {
             self.limit = Some(value as usize);
         }
     }
 
     fn record_str(&mut self, field: &Field, value: &str) {
```

Comment:
It seems `record_str` is only invoked for fields that are `&str`.  If you changed the `by_key.rs` example to be:

```rust
            for key in &[String::from("foo"), String::from("bar")] {
             ....
                    internal_log_rate_key = %key,
```

then they won't be used as a rate limit key. If we also override `record_debug` and duplicate this functionality, then it would work.


Without `record_debug` you would still be able to get it working by coercing the string to a str like:

```rust
            for key in &[String::from("foo"), String::from("bar")] {
             ....
                    internal_log_rate_key = key.as_str(),
```

But, I bet people would always forget this, and there is no way for the code to alert that this has happened...

Reply:
It looks like you can't actually pass a `String` there, it has to be a `tracing::Value`: https://docs.rs/tracing/0.1.10/tracing/trait.Value.html

Reply:
This does limit the usage to string values though (as opposed to, for example, an integer. I think I may actually back out this particular change though as I think the implementation that rate limits just on span context fields should be sufficient for both of our current needs: rate limiting in remap and sharing internal events.

Reply:
> It looks like you can't actually pass a `String` there,

You can pass a String - or any other type, but in the macro you prepend it with `%`, which essentially takes the `Display` value. Prepending with `?` takes the `Debug` value. 

It seems that there is no `record_display` method, not sure why, - both types go through the `record_debug` call. The default implementation for the other types just calls `record_debug`.

Not sure I understand what you mean here?

> This does limit the usage to string values

Reply:
Aha, I missed your introduction of the `%`. Yeah, I see.

For

> This does limit the usage to string values

I just mean that the visitor that records the value for the key is currently only looking at `&str`s via `visit_str` so if, for example, a `uint` was passed, it wouldn't pick that up.

I still think I'm going to back out this specific change of allowing a key, in-lieu of using tracing span data, but I'll carry over your suggestion to the tracing span visitor.

I did want to check with you to see if just using span data would be sufficient for the remap case though. I was thinking that remap could add a span that records the source line number. In addition to being able to use this to group the log messages so they are rate limited separately, I also thought this would be a useful enhancement to have internal vector logs also be tagged with the remap source line number to aid with debugging. Maybe we could pair on this this week?

Reply:
I implemented `record_debug` for the `message` field. I also discussed with Stephen offline that the span approach should be sufficient for VRL too so I dropped the implementation of `internal_rate_limit_key`.